#128 migrate project to pyrpoject toml based package setup

Nicoretti opened the pull request:
fixes #128

Nicoretti committed:
Add pyproject.toml

Nicoretti committed:
Add basic project information to pyproject.toml

Nicoretti committed:
Add dependencies to pyproject.toml

Nicoretti committed:
Remove deprecated and unused configuration entries

Nicoretti committed:
Update include list in pyproject.toml

Nicoretti committed:
Add extension points required for sqlalchemy plugin(s)

Nicoretti committed:
Remove unused script

Nicoretti committed:
Add sync support for project.toml based version

kaklakariada commented on scripts/version_check.py:
```diff
@@ -0,0 +1,134 @@
+# fmt: off
+_VERSION_MODULE_TEMPLATE = cleandoc('''
+# ATTENTION: 
+# This file is generated, do not edit this file manually!
```
```suggestion
# This file is generated, do not edit it manually!
```

Nicoretti replied:
Done

kaklakariada commented on scripts/version_check.py:
```diff
@@ -0,0 +1,134 @@
+_VERSION_MODULE_TEMPLATE = cleandoc('''
+# ATTENTION: 
+# This file is generated, do not edit this file manually!
+# If you need to change the version, do so in the project.toml, e.g. by using `poetry version X.Y.Z`.
```
👍

kaklakariada commented on sqlalchemy_exasol/version.py:
```diff
@@ -0,0 +1,9 @@
+# ATTENTION: 
+# This file is generated, do not edit this file manually!
+# If you need to change the version, do so in the project.toml, e.g. by using `poetry version X.Y.Z`.
```
Do we need to check in this generated file or is it possible to create it at build time? Or does the pre-commit-hook do this automatically?

Nicoretti replied:
Very good point! I have thought about this quite a bit also because the context is quite complex.
The TL;DR more or less is this:
* It is not as straight forward to do it like that as within the old "build system"
* It can be done with package metadata though see [this comment](https://github.com/python-poetry/poetry/pull/2366#issuecomment-652418094)
* Using metadata prior to `python 3.8` has quite some pitfalls
  ("Dependency hell of the metadata-lib backport")
  
So my plan was: for now add a git hook and verification step to the build and when the library is [upgraded to 3.8](https://github.com/exasol/sqlalchemy-exasol/issues/135#issuecomment-1143186610) switch to the metadatalib variant.

tkilias replied:
Regarding the upgrade to python 3.8, I think this project can wait longer with that, we only need to drop 3.6 if didn't do it yet. And, please don't mix things in issues together. The version file is a separate issue. I guess you want to use importlib.metadata, for pre 3.8 there is a Backport package, you can use. We already used it already various other projects.

tkilias replied:
However, there is another issue with importlin_metadata. From what we saw, it only works for installed packages.

Nicoretti replied:
@tkilias that it works only on installed package is fine in the scenario of the sqlaclhemy plugin I would say (I don't expect the extension points to work without installation, I am not 100% sure though).

Regarding `importlib.metadata` previous to 3.8 where it is part of the standard library can be hell of a mess if more than one library uses it, which is often the case (We don't want to run into dependency issues because of the version imho).

tkilias commented on setup.cfg:
```diff
@@ -1,6 +1,3 @@
-[egg_info]
-#tag_build = dev
-
 [tool:pytest]
```
I think, this can move as well https://docs.pytest.org/en/6.2.x/customize.html#configuration-file-formats

Nicoretti replied:
@tkilias sadly no, I tried (for details see https://github.com/exasol/sqlalchemy-exasol/issues/144)

tkilias replied:
Ok, good to know, than I suggest add a comment

Nicoretti replied:
Good point, I will do so.

Nicoretti committed:
Add version check to ci/verifier action

Nicoretti committed:
Add justification why the pytest config still is part of setup.cfg

Nicoretti committed:
Remove condaforge reference in README

Conda forge is not maintained anymore.

Nicoretti committed:
Fix error in setup.cfg

Nicoretti committed:
Replace version information mechansim

For now the version and iformation will be updated using a git hook and nox target.
When the project is upgraded to 3.8 we should replace this mechanism by
the use of importlib.metadata.

see also:
* https://github.com/exasol/sqlalchemy-exasol/pull/145#discussion_r885733975
* https://github.com/python-poetry/poetry/pull/2366#issuecomment-652418094
* https://github.com/exasol/sqlalchemy-exasol/issues/135#issuecomment-1143186610

Nicoretti committed:
Add poetry lock file

Nicoretti committed:
Delete deprecated requirements*.txt files

Nicoretti committed:
Remove deprecated setup.py

Nicoretti committed:
Update comment in generated file according to code review

Nicoretti committed:
Update CI workflow to use poetry

Nicoretti committed:
Fix command invocation to use poetry

Nicoretti committed:
Fix import of version module

Nicoretti committed:
Fix installation order

Nicoretti committed:
Use python built in mock library

Nicoretti committed:
Remove deprecated version parsing

Nicoretti committed:
Add missing newline

Nicoretti committed:
Rename workflow step

kaklakariada commented on .github/workflows/CI.yml:
```diff
@@ -35,26 +35,24 @@ jobs:
     - name: Setup Python ${{ matrix.python }}
       uses: actions/setup-python@v2
       with:
         python-version: ${{ matrix.python }}
```
```suggestion
    - name: Setup Python ${{ matrix.python }}
      uses: actions/setup-python@v3
      with:
        python-version: ${{ matrix.python }}
        cache: 'poetry'
```

see https://github.com/actions/setup-python#caching-packages-dependencies

kaklakariada commented on .github/workflows/CI.yml:
```diff
@@ -35,26 +35,24 @@ jobs:
     - name: Setup integration-test-docker-environment
       uses: actions/setup-python@v2
 
```
```suggestion
```

is the setup-python action duplicated?

kaklakariada commented on .github/workflows/CI.yml:
```diff
@@ -94,11 +90,10 @@ jobs:
       with:
         python-version: ${{ matrix.python }}
```
Also use version v3 and activate caching for poetry.

tkilias commented on .github/workflows/CI.yml:
```diff
@@ -65,21 +63,19 @@ jobs:
 
     - name: Check documentation links ${{ matrix.python }} using ${{ matrix.connector }}
-      run: nox -s "check-links"
+      run: poetry run nox -s "check-links"
```
you might want to move this to its own workflow, because check links can be flaky and you don't want to prevent the test running because some links don't respond properly

tkilias commented on .pre-commit-config.yaml:
```diff
@@ -0,0 +1,11 @@
+      always_run: true
+      verbose: true
+      language: system
+      entry: python scripts/version_check.py sqlalchemy_exasol/version.py
```
Theoretically, you could simply use the nox command check_version

Nicoretti replied:
Yes, but for now I wanted nox to check and for the hook I wanted it to fix it if necessary.

tkilias commented on sqlalchemy_exasol/pyodbc.py:
```diff
@@ -9,7 +9,7 @@
 import sys
 import logging
-from distutils.version import LooseVersion
+from packaging import version
```
is this an additional package, if yes then add it as explicit dependency, because I think, it is not listed, yet. However, I like that we replace distutils

Nicoretti replied:
Good catch, thx! Did not list it explicitly!

Nicoretti committed:
Update .github/workflows/CI.yml

Co-authored-by: chπ <christoph.pirkl@exasol.com>

Nicoretti committed:
Update .github/workflows/CI.yml

Co-authored-by: chπ <christoph.pirkl@exasol.com>

Nicoretti committed:
Add check-links workflow

Nicoretti committed:
Add packaging dependency explicitly

Nicoretti committed:
Update ci workflow to support caching

Nicoretti committed:
Fix github workflows to install poetry previous to the python setup

If poetry is not installed before the python action gets installed/run
it will fail if caching is enabled.

Nicoretti committed:
Fix check-links workflow

Nicoretti committed:
Renamed workflow check-links to link-check

Nicoretti committed:
Fix issue of wrong selected python version in github action

Nicoretti committed:
Rework CI-CD workflow to use poetry

Nicoretti committed:
Remove version file from ignore list

Nicoretti committed:
Remove MANIFEST.in file

Nicoretti committed:
Add dependabot updater to repository

Nicoretti committed:
Update base python version to 3.8

Nicoretti committed:
Add pyupgrade commit hook to workspace

Nicoretti committed:
Update python version(s) in github actions

Nicoretti committed:
Bump version to 3.0.0

Nicoretti committed:
Removed verbose flag from pre commit hook

Nicoretti committed:
Show CI instead of CI-CD status in README

Nicoretti committed:
Update information about minimum python version

Nicoretti committed:
Update changelog

Fixes #135

Nicoretti committed:
Update minimum required python version in developer guide

Nicoretti committed:
Update developer guide

tkilias commented on .github/workflows/ci.yaml:
```diff
@@ -65,21 +63,19 @@ jobs:
 
     - name: Check documentation links ${{ matrix.python }} using ${{ matrix.connector }}
-      run: nox -s "check-links"
+      run: poetry run nox -s "check-links"
```
You run this already in an extra workflow, you can remove this, such that we don't fail the test, because some links don't work

Nicoretti replied:
good catch thx!

tkilias commented on pyproject.toml:
```diff
@@ -0,0 +1,77 @@
+SQLAlchemy = ">=1.3.24,<1.4"
+pyodbc = ">=4.0.32"
+packaging = "^21.3"
+turbodbc = {path = "../../Downloads/turbodbc-4.5.3.tar.gz"}
```
this needs to be changed later

Nicoretti replied:
:white_check_mark:  agreed, currently I need it because I wait on the turbodbc fix.
(will be removed/changed in the final version)

Nicoretti committed:
Add todo for updating turbodbc dependency

Nicoretti committed:
Remove link check from test workflow

Nicoretti committed:
Update version check hook and script

Nicoretti committed:
Remove link check from CI - CD workflow

Nicoretti committed:
Add git script/module to query available tags

Nicoretti committed:
Add release support for new project format to noxfile and CI/CD

Nicoretti committed:
Prepare 3.0.0 release

Nicoretti committed:
Fix missing check in verify step

tkilias commented on noxfile.py:
```diff
@@ -83,20 +86,57 @@ def temporary_odbc_config(config):
+    module_version = version_from_poetry()
+    git_version = version_from_string(tags()[-1])
+
+    if not (module_version == git_version == version_file):
```
Are you sure this is correct?

Nicoretti replied:
I think so, should be only triggered on tagging/tag push

tkilias commented on noxfile.py:
```diff
@@ -83,20 +86,57 @@ def temporary_odbc_config(config):
+    if cli_args.username:
+        args.append("--username")
+        args.append(cli_args.username)
+    if cli_args.password:
```
Do we need a interactive password request?

Nicoretti replied:
Yes would make sense, at least when you run it locally, generally it's not recommended to have the pw within a cli call.

Nicoretti committed:
Make sure no unnecessary virtual environments will be created

tkilias commented on CHANGES.md:
```diff
@@ -1,3 +1,11 @@
+* Dropped python 3.7 support
+  * If you still depend on python 3.7 use the 2.x version line
+* Updated dependencies, most noticeably
+  * turbodbc to 4.x.y
```
Hm, is this still valid

tkilias commented on .github/workflows/ci-cd.yml:
```diff
@@ -91,17 +92,10 @@ jobs:
+        poetry-version: 1.1.13
+
+    - name: Build and push package to PYPI
+      run: poetry run nox -s release -- --username=__token__ --password=${{ secrets.pypi_token }}
```
Is this best practices to Adresse the secret this way

Nicoretti replied:
I removed the password flag, I can see that it could be used locally by accident. Switched to env var only set during exactly the publish invocation. Poetry [config](https://python-poetry.org/docs/repositories/#configuring-credentials) supports keyring, but not sure if that's in the image out of the box and if it fails it will fall back to write it to a file.

tkilias replied:
That is actually better, because we usually only want to release from the CI

tkilias commented on pyproject.toml:
```diff
@@ -0,0 +1,80 @@
+    "Programming Language :: Python :: 3",
+    "Programming Language :: Python :: 3.7",
+    "Programming Language :: Python :: 3.8",
+    "Programming Language :: Python :: 3.9",
```
3.10 is missing

Nicoretti replied:
Good catch!

Nicoretti committed:
Update pytest dependency

Nicoretti committed:
Address pytest compatibility issue

tkilias commented on odbcconfig/odbcinst.ini:
```diff
@@ -4,4 +4,4 @@
 [EXAODBC]
 #Driver location will be appended in build environment:
-#DRIVER=<PATH TO DRIVER>
+DRIVER=/home/nic/Projects/sqlalchemy-exasol/driver/libexaodbc-uo2214lv1.so
```
Local path. Should we provide a template file instead and add this file to gitignore

Nicoretti replied:
Good point, did not happen the first time!

Nicoretti committed:
Adjust release task and action

Nicoretti committed:
Remove odbcinst.ini and replace with template

tkilias commented on .github/workflows/ci-cd.yml:
```diff
@@ -98,4 +98,4 @@ jobs:
 
     - name: Build and push package to PYPI
-      run: poetry run nox -s release -- --username=__token__ --password=${{ secrets.pypi_token }}
+      run: POETRY_PYPI_TOKEN_PYPI="${{ secrets.pypi_token }}" poetry run nox -s release
```
Use env property of GitHub action

tkilias replied:
Can we try this, I saw some issues in the poetry repos for the env variable

Nicoretti replied:
Can you elaborate what you mean exactly?

tkilias commented on odbcconfig/odbcinst.ini.template:
```diff
@@ -0,0 +1,7 @@
+
+[EXAODBC]
+#Driver location needs to be added in build environment:
+DRIVER=<PATH_TO_DRIVER>
```
This template is used by a user and not a program or? If it would be used by a program, we should use Jinja or so

Nicoretti replied:
User

Nicoretti committed:
Harden relase mechanism

Either use environment variables for credentials if available,
otherwise to fallback to interactivly asking for the credentials.
